src/main/java/com/linkedin/kmf/services/TopicManagementServiceManager.java

```diff
@@ -0,0 +1,154 @@
+    Map<String, Object> topicManagementServiceProps = props.containsKey(MirrorPipelineMonitorConfig.TOPIC_MANAGEMENT_SERVICE_CONFIG)
+      ? (Map) props.get(MirrorPipelineMonitorConfig.TOPIC_MANAGEMENT_SERVICE_CONFIG) : new HashMap<>();
+    topicManagementServiceProps.put(CommonServiceConfig.TOPIC_CONFIG, topic);
+    return topicManagementServiceProps;
+  }
+
+
+  public Map<String, Integer> getBrokerCounts(List<String> zookeeperUrls) {
+    Map<String, Integer> zookeeperBrokerCount = new HashMap<>();
+    for (String zookeeper: zookeeperUrls) {
+      if (zookeeperBrokerCount.containsKey(zookeeper)) {
+        throw new RuntimeException(_name + ": Duplicate zookeeper connections specified for topic management.");
+      }
+
+      int brokerCount = com.linkedin.kmf.common.Utils.getBrokerCount(zookeeper);
+      zookeeperBrokerCount.put(zookeeper, brokerCount);
+    }
+
+    return zookeeperBrokerCount;
+  }
+
+  public int getMaxNumberOfPartitions(Map<String, Integer> brokerCounts) {
+    int maxNumPartitions = 0;
+    for (String zookeeper: brokerCounts.keySet()) {
+      int brokerCount = brokerCounts.get(zookeeper);
+      int numPartitions = (int) Math.ceil(_partitionsToBrokerRatio * brokerCount);
+      maxNumPartitions = Math.max(maxNumPartitions, numPartitions);
+    }
+
+    return maxNumPartitions;
+  }
+
+  @Override
+  public void start() {
+    if (_running.compareAndSet(false, true)) {
+      for (TopicManagementService service : _services.values()) {
+        service.start();
+      }
+
+      _managementExecutor.scheduleWithFixedDelay(new TopicManagementServiceManagerRunnable(), 40_000, 40_000, TimeUnit.MILLISECONDS);
+      LOG.info(_name + "/TopicManagementServiceManager started");
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (_running.compareAndSet(true, false)) {
+      _managementExecutor.shutdown();
+      for (TopicManagementService service: _services.values()) {
+        service.stop();
+      }
+      LOG.info(_name + "/TopicManagementServiceManager stopped");
+    }
+  }
+
+  @Override
+  public boolean isRunning() {
+    return _running.get();
+  }
+
+  @Override
+  public void awaitShutdown() {
+    try {
+      _managementExecutor.awaitTermination(Integer.MAX_VALUE, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      Thread.interrupted();
+    }
+    LOG.info(_name + "/TopicManagementServiceManager shutdown completed");
+  }
+
+  private class TopicManagementServiceManagerRunnable implements Runnable {
+
+    @Override
+    public void run() {
+      Map<String, Integer> brokerCounts = getBrokerCounts(_zookeeperUrls);
+      int maxNumPartitions = getMaxNumberOfPartitions(brokerCounts);
+
+      for (String zookeeper: _zookeeperUrls) {
+        double individualPartitionToBrokerRatio = maxNumPartitions / brokerCounts.get(zookeeper);
+        _services.get(zookeeper).setPartitionsToBrokerRatio(individualPartitionToBrokerRatio);
```

Comment:
I have a few concern with this approach:

- This doesn't necessarily ensure that the partition count will be maxNumPartitions in all clusters. TopicManagementService only expand partition when PARTITIONS_TO_BROKER_RATIO_THRESHOLD is violated.

- It seems hacky to change the config of another class (i.e. TopicManagementService) on the fly. This makes it harder to reason about TopicManagementService and extend it in the future. We should try to avoid this.

- Currently the default value of REBALANCE_INTERVAL_MS_CONFIG is 10 min. It means that the partition count may be inconsistent for up to 10 minutes since TopicMangementService of each cluster runs at different time. Reducing this interval may alleviate the problem but it is not a good idea to rely on a low interval to reduce race condition. Ideally we want a single thread to change partition count in all clusters altogether.